File: app/src/main/java/com/itcuties/android/apps/MainActivity.java
package com.itcuties.android.apps;

import java.util.ArrayList;
import java.util.List;

import android.app.ListActivity;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.widget.ListView;
import android.widget.Toast;

import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

import com.itcuties.android.apps.data.SMSData;

public class MainActivity extends ListActivity {

    List<SMSData> smsList = new ArrayList<SMSData>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        final int REQUEST_CODE_ASK_PERMISSIONS = 123;
        ActivityCompat.requestPermissions(MainActivity.this, new String[]{"android.permission.READ_SMS"}, REQUEST_CODE_ASK_PERMISSIONS);

        if (ContextCompat.checkSelfPermission(getBaseContext(), "android.permission.READ_SMS") == PackageManager.PERMISSION_GRANTED) {
            Uri uri = Uri.parse("content://sms/inbox");
            Cursor c = getContentResolver().query(uri, null, null, null, null);
            startManagingCursor(c);

            // Read the sms data and store it in the list
            if (c != null) {
                if (c.moveToFirst()) {
                    for (int i = 0; i < c.getCount(); i++) {
                        SMSData sms = new SMSData();
                        sms.setBody(c.getString(c.getColumnIndexOrThrow("body")));
                        sms.setNumber(c.getString(c.getColumnIndexOrThrow("address")));
                        smsList.add(sms);

                        c.moveToNext();
                    }
                }
                c.close();
            }
            // Set smsList in the ListAdapter
            setListAdapter(new ListAdapter(this, smsList));
        }

    }

}

File: app/src/main/java/com/itcuties/android/apps/MainActivity.kt
package com.itcuties.android.apps

import android.content.pm.PackageManager
import android.database.Cursor
import android.net.Uri
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import kotlinx.android.synthetic.main.activity_main.*
import java.lang.Exception

class MainActivity : AppCompatActivity() {

    private val requestCode = 123
    private var cursor: Cursor? = null
    private lateinit var mainViewModel: MainViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        ActivityCompat.requestPermissions(this, arrayOf("android.permission.READ_SMS"), requestCode)

        mainViewModel = ViewModelProvider(this).get(MainViewModel::class.java)

        if (ContextCompat.checkSelfPermission(baseContext, "android.permission.READ_SMS") == PackageManager.PERMISSION_GRANTED) {
            val uri = Uri.parse("content://sms/inbox")
            try {
                cursor = contentResolver.query(uri, null, null, null, null)
                cursor?.let {
                    startManagingCursor(cursor)
                    mainViewModel.getSmsList(cursor)
                }
            } catch (ignore: Exception) {

            } finally {
                cursor?.close()
            }

            mainViewModel.smsListLD.observe(this, Observer {

                it?.let {
                    cursor?.close()
                    list.adapter = ListAdapter(this, it)
                }
            })

        }

    }

}
